core/ledger/kvledger/unjoin_channel_test.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package kvledger
+
+import (
+	"testing"
+
+	configtxtest "github.com/hyperledger/fabric/common/configtx/test"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/msgs"
+	"github.com/hyperledger/fabric/core/ledger/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestUnjoinChannel(t *testing.T) {
+	conf, cleanup := testConfig(t)
+	conf.HistoryDBConfig.Enabled = false
+	defer cleanup()
+
+	ledgerID := "ledger_unjoin"
+
+	provider := testutilNewProvider(conf, t, &mock.DeployedChaincodeInfoProvider{})
+	activeLedgerIDs, err := provider.List()
+	require.NoError(t, err)
+	require.Len(t, activeLedgerIDs, 0)
+
+	genesisBlock, err := configtxtest.MakeGenesisBlock(ledgerID)
+	require.NoError(t, err)
+	_, err = provider.CreateFromGenesisBlock(genesisBlock)
+	require.NoError(t, err)
+
+	activeLedgerIDs, err = provider.List()
+	require.NoError(t, err)
+	require.Len(t, activeLedgerIDs, 1)
+	provider.Close()
+
+	// Unjoin the channel from the peer
+	err = UnjoinChannel(conf, ledgerID)
+	require.NoError(t, err)
+
+	// channel should no longer be present in the channel list
+	provider = testutilNewProvider(conf, t, &mock.DeployedChaincodeInfoProvider{})
+	activeLedgerIDs, err = provider.List()
+	require.NoError(t, err)
+	require.Len(t, activeLedgerIDs, 0)
+
+	provider.Close()
+}
+
+// Unjoining an unjoined channel is OK.
+func TestUnjoinUnjoinedChannel(t *testing.T) {
+	conf, cleanup := testConfig(t)
+	conf.HistoryDBConfig.Enabled = false
+	defer cleanup()
+
+	ledgerID := "ledger_unjoin_unjoined"
+
+	provider := testutilNewProvider(conf, t, &mock.DeployedChaincodeInfoProvider{})
+	genesisBlock, err := configtxtest.MakeGenesisBlock(ledgerID)
+	require.NoError(t, err)
+	_, err = provider.CreateFromGenesisBlock(genesisBlock)
+	require.NoError(t, err)
+	provider.Close()
+
+	// unjoin the channel
+	require.NoError(t, UnjoinChannel(conf, ledgerID))
+
+	provider = testutilNewProvider(conf, t, &mock.DeployedChaincodeInfoProvider{})
+	activeLedgerIDs, err := provider.List()
+	require.NoError(t, err)
+	require.Len(t, activeLedgerIDs, 0)
+	provider.Close()
+
+	// Subsequent unjoins will not throw an error
+	require.NoError(t, UnjoinChannel(conf, ledgerID))
+	require.NoError(t, UnjoinChannel(conf, ledgerID))
+	require.NoError(t, UnjoinChannel(conf, ledgerID))
```

Comment:
- This is not a desired behavior. You got this here because you missed deleting the entry from the idStore. Otherwise, it would have rightly returned this [error](https://github.com/hyperledger/fabric/blob/4c4e58c1228968cefc9d5c027f10884fb3707174/core/ledger/kvledger/kv_ledger_provider.go#L634).

- Another aspect of the behavior would be that it should be able to create the ledger with the same ID again. But this test perhaps fits better in kvledger/tests folder - where we maintain the tests for testing the behavior under various scenarios (including when the ledger contains private data) unlike in this package where the tests are more for very basic code coverage. Keeping that in mind, I would suggest that the two tests that you may want to retain here in package would be the first and last only as the rest of the tests would be covered in kvledger/tests package.